File: frontend/src/lib/environmentalImpact.ts
export interface TokenUsageJson {
  prompt_tokens?: number
  completion_tokens?: number
  total_tokens?: number
  image_count?: number
  estimated_tokens?: boolean
  environmental_impact?: EnvironmentalImpactMetrics
}

export interface EnvironmentalImpactMetrics {
  energy_wh: number
  co2_grams: number
  water_ml: number
  water_drops: number
  prompt_tokens?: number
  completion_tokens?: number
  total_tokens?: number
  image_count?: number
  request_equivalent?: number
  model_intensity?: number
}

export interface EnvironmentalFootprintResponse {
  totals: EnvironmentalImpactMetrics & {
    interaction_count: number
    estimated_entry_count: number
  }
  actor_type: 'teacher' | 'student'
}

const BASELINE_ENERGY_WH = 0.24
const BASELINE_CO2_GRAMS = 0.03
const BASELINE_WATER_ML = 0.26
const BASELINE_WATER_DROPS = 5
const INPUT_TOKEN_WEIGHT = 0.35
const OUTPUT_TOKEN_WEIGHT = 1.0
const BASELINE_WEIGHTED_TOKENS = 1600

const MODEL_INTENSITY_OVERRIDES: Record<string, number> = {
  'gpt-5-nano': 0.7,
  'gpt-5-mini': 1.0,
  'gpt-4o-mini': 0.9,
  'gpt-4o': 1.4,
  'claude-haiku-4-5-20251001': 1.05,
  'claude-3-haiku-20240307': 0.95,
  'claude-3-5-sonnet-20241022': 1.35,
  'gemini-2.0-flash-lite': 0.8,
  'gemini-2.0-flash': 0.9,
  'gemini-3.1-flash-lite-preview': 0.82,
  'mistral-nemo': 0.95,
  mistral: 1.0,
  'deepseek-chat': 0.9,
  'deepseek-reasoner': 1.3,
}

const PROVIDER_INTENSITY_FALLBACK: Record<string, number> = {
  openai: 1.0,
  anthropic: 1.15,
  gemini: 0.9,
  deepseek: 0.92,
  ollama: 0.98,
  flux: 1.6,
  system: 0,
  fallback: 1.0,
}

const IMAGE_REQUEST_EQUIVALENT: Record<string, number> = {
  'dall-e-3': 4.5,
  'gpt-image-1': 4.8,
  'flux-schnell': 2.6,
  'flux-dev': 3.0,
  sdxl: 2.9,
}

function modelIntensity(provider?: string, model?: string) {
  const normalizedModel = (model || '').trim()
  const normalizedProvider = (provider || '').trim().toLowerCase()

  if (MODEL_INTENSITY_OVERRIDES[normalizedModel] != null) {
    return MODEL_INTENSITY_OVERRIDES[normalizedModel]
  }

  const prefixHit = Object.entries(MODEL_INTENSITY_OVERRIDES).find(([key]) => normalizedModel.startsWith(key))
  if (prefixHit) return prefixHit[1]
  if (normalizedModel.startsWith('gpt-5')) return 1.05
  if (normalizedModel.startsWith('gpt-4')) return 1.35
  if (normalizedModel.startsWith('claude')) return 1.2
  if (normalizedModel.startsWith('gemini')) return 0.9

  return PROVIDER_INTENSITY_FALLBACK[normalizedProvider] ?? 1
}

export function estimateEnvironmentalImpact(
  tokenUsage?: TokenUsageJson | null,
  provider?: string | null,
  model?: string | null,
): EnvironmentalImpactMetrics | null {
  if (tokenUsage?.environmental_impact) {
    return tokenUsage.environmental_impact
  }

  const promptTokens = Math.max(0, Number(tokenUsage?.prompt_tokens || 0))
  const completionTokens = Math.max(0, Number(tokenUsage?.completion_tokens || 0))
  const imageCount = Math.max(0, Number(tokenUsage?.image_count || 0))
  const hasData = promptTokens > 0 || completionTokens > 0 || imageCount > 0 || provider || model
  if (!hasData) return null

  const weightedTokens = promptTokens * INPUT_TOKEN_WEIGHT + completionTokens * OUTPUT_TOKEN_WEIGHT
  const tokenRequestEquivalent = weightedTokens / BASELINE_WEIGHTED_TOKENS
  const imageEquivalent = imageCount > 0 ? (IMAGE_REQUEST_EQUIVALENT[(model || '').trim()] ?? 3.0) * imageCount : 0
  const requestEquivalent = Math.max(imageEquivalent, tokenRequestEquivalent, weightedTokens === 0 && imageCount === 0 ? 1 : 0)
  const intensity = modelIntensity(provider || undefined, model || undefined)
  const impactFactor = requestEquivalent * intensity
  const waterMl = BASELINE_WATER_ML * impactFactor

  return {
    energy_wh: Number((BASELINE_ENERGY_WH * impactFactor).toFixed(4)),
    co2_grams: Number((BASELINE_CO2_GRAMS * impactFactor).toFixed(4)),
    water_ml: Number(waterMl.toFixed(4)),
    water_drops: Math.max(0, Math.round((waterMl / BASELINE_WATER_ML) * BASELINE_WATER_DROPS)),
    prompt_tokens: promptTokens,
    completion_tokens: completionTokens,
    total_tokens: promptTokens + completionTokens,
    image_count: imageCount,
    request_equivalent: Number(requestEquivalent.toFixed(4)),
    model_intensity: Number(intensity.toFixed(4)),
  }
}

File: backend/app/services/environmental_impact.py
from __future__ import annotations

from math import ceil, sqrt
from typing import Any, Optional


BASELINE_ENERGY_WH = 0.24
BASELINE_CO2_GRAMS = 0.03
BASELINE_WATER_ML = 0.26
BASELINE_WATER_DROPS = 5

INPUT_TOKEN_WEIGHT = 0.35
OUTPUT_TOKEN_WEIGHT = 1.0
BASELINE_WEIGHTED_TOKENS = 1600.0

MODEL_INTENSITY_OVERRIDES = {
    "gpt-5-nano": 0.7,
    "gpt-5-mini": 1.0,
    "gpt-4o-mini": 0.9,
    "gpt-4o": 1.4,
    "claude-haiku-4-5-20251001": 1.05,
    "claude-3-haiku-20240307": 0.95,
    "claude-3-5-sonnet-20241022": 1.35,
    "gemini-2.0-flash-lite": 0.8,
    "gemini-2.0-flash": 0.9,
    "gemini-3.1-flash-lite-preview": 0.82,
    "mistral-nemo": 0.95,
    "mistral": 1.0,
    "deepseek-chat": 0.9,
    "deepseek-reasoner": 1.3,
}

PROVIDER_INTENSITY_FALLBACK = {
    "openai": 1.0,
    "anthropic": 1.15,
    "gemini": 0.9,
    "deepseek": 0.92,
    "ollama": 0.98,
    "flux": 1.6,
    "system": 0.0,
    "fallback": 1.0,
}

IMAGE_REQUEST_EQUIVALENT = {
    "dall-e-3": 4.5,
    "gpt-image-1": 4.8,
    "flux-schnell": 2.6,
    "flux-dev": 3.0,
    "sdxl": 2.9,
}


def estimate_tokens_from_text(text: Optional[str]) -> int:
    if not text:
        return 0
    normalized = " ".join(str(text).split())
    if not normalized:
        return 0
    return max(1, ceil(len(normalized) / 4))


def estimate_prompt_tokens_from_messages(messages: list[dict[str, Any]]) -> int:
    return sum(estimate_tokens_from_text(m.get("content")) for m in messages)


def build_estimated_token_usage(
    messages: list[dict[str, Any]],
    response_content: Optional[str],
    image_count: int = 0,
) -> dict[str, Any]:
    prompt_tokens = estimate_prompt_tokens_from_messages(messages)
    completion_tokens = estimate_tokens_from_text(response_content)
    return {
        "prompt_tokens": prompt_tokens,
        "completion_tokens": completion_tokens,
        "total_tokens": prompt_tokens + completion_tokens,
        "image_count": max(0, int(image_count or 0)),
        "estimated_tokens": True,
    }


def _model_intensity(provider: Optional[str], model: Optional[str]) -> float:
    normalized_model = (model or "").strip()
    normalized_provider = (provider or "").strip().lower()

    if normalized_model in MODEL_INTENSITY_OVERRIDES:
        return MODEL_INTENSITY_OVERRIDES[normalized_model]

    for known_model, factor in MODEL_INTENSITY_OVERRIDES.items():
        if normalized_model.startswith(known_model):
            return factor

    if normalized_model.startswith("gpt-5"):
        return 1.05
    if normalized_model.startswith("gpt-4"):
        return 1.35
    if normalized_model.startswith("claude"):
        return 1.2
    if normalized_model.startswith("gemini"):
        return 0.9

    return PROVIDER_INTENSITY_FALLBACK.get(normalized_provider, 1.0)


def _image_request_equivalent(model: Optional[str], image_count: int) -> float:
    if image_count <= 0:
        return 0.0
    normalized_model = (model or "").strip()
    factor = IMAGE_REQUEST_EQUIVALENT.get(normalized_model, 3.0)
    return factor * image_count


def calculate_environmental_impact(
    *,
    provider: Optional[str],
    model: Optional[str],
    prompt_tokens: int = 0,
    completion_tokens: int = 0,
    image_count: int = 0,
    request_count_fallback: int = 1,
) -> dict[str, Any]:
    safe_prompt_tokens = max(0, int(prompt_tokens or 0))
    safe_completion_tokens = max(0, int(completion_tokens or 0))
    safe_image_count = max(0, int(image_count or 0))
    weighted_tokens = (
        safe_prompt_tokens * INPUT_TOKEN_WEIGHT
        + safe_completion_tokens * OUTPUT_TOKEN_WEIGHT
    )
    token_request_equivalent = weighted_tokens / BASELINE_WEIGHTED_TOKENS
    request_equivalent = max(
        _image_request_equivalent(model, safe_image_count),
        token_request_equivalent,
        float(request_count_fallback if weighted_tokens == 0 and safe_image_count == 0 else 0),
    )
    intensity = _model_intensity(provider, model)
    impact_factor = request_equivalent * intensity

    energy_wh = BASELINE_ENERGY_WH * impact_factor
    co2_grams = BASELINE_CO2_GRAMS * impact_factor
    water_ml = BASELINE_WATER_ML * impact_factor

    return {
        "energy_wh": round(energy_wh, 4),
        "co2_grams": round(co2_grams, 4),
        "water_ml": round(water_ml, 4),
        "water_drops": max(0, round(water_ml / BASELINE_WATER_ML * BASELINE_WATER_DROPS)),
        "prompt_tokens": safe_prompt_tokens,
        "completion_tokens": safe_completion_tokens,
        "total_tokens": safe_prompt_tokens + safe_completion_tokens,
        "image_count": safe_image_count,
        "request_equivalent": round(request_equivalent, 4),
        "model_intensity": round(intensity, 4),
    }


def enrich_usage_with_environmental_impact(
    usage: Optional[dict[str, Any]],
    *,
    provider: Optional[str],
    model: Optional[str],
    request_count_fallback: int = 1,
) -> dict[str, Any]:
    base_usage = dict(usage or {})
    impact = calculate_environmental_impact(
        provider=provider,
        model=model,
        prompt_tokens=int(base_usage.get("prompt_tokens") or 0),
        completion_tokens=int(base_usage.get("completion_tokens") or 0),
        image_count=int(base_usage.get("image_count") or 0),
        request_count_fallback=request_count_fallback,
    )
    base_usage.setdefault("total_tokens", impact["total_tokens"])
    base_usage["environmental_impact"] = impact
    return base_usage
